lastgersync/evmdownloader.go

```diff
@@ -128,50 +136,52 @@ func (d *downloader) Download(ctx context.Context, fromBlock uint64, downloadedC
 			break
 		}
 
-		blockHeader, isCanceled := d.GetBlockHeader(ctx, fromBlock)
+		header, isCanceled := d.GetBlockHeader(ctx, fromBlock)
 		if isCanceled {
 			return
 		}
 
 		block := &sync.EVMBlock{
 			EVMBlockHeader: sync.EVMBlockHeader{
-				Num:        blockHeader.Num,
-				Hash:       blockHeader.Hash,
-				ParentHash: blockHeader.ParentHash,
-				Timestamp:  blockHeader.Timestamp,
+				Num:        header.Num,
+				Hash:       header.Hash,
+				ParentHash: header.ParentHash,
+				Timestamp:  header.Timestamp,
 			},
 		}
-		// Set the greatest GER injected from the list
-		d.setGreatestGERInjectedFromList(block, gers)
+		// Set the greatest GER injected from retrieved GERs
+		d.populateGreatestInjectedGER(block, gers)
 
-		downloadedCh <- *block
-		// Update nextIndex based on the last injected GER event
+		// Update nextIndex based on the last injected GER info
 		if len(block.Events) > 0 {
-			event, ok := block.Events[0].(Event)
-			if !ok {
-				log.Errorf("unexpected type %T in events", block.Events[0])
+			if e, ok := block.Events[0].(*GlobalExitRootInfo); ok {
+				nextL1InfoTreeIndex = e.L1InfoTreeIndex + 1
+			}
+			downloadedCh <- *block
+		} else {
+			for _, block := range d.GetEventsByBlockRange(ctx, fromBlock, fromBlock) {
```

Comment:
Hmm, this looks a bit complex. You try to read it from the GER map, and if that doesn't help, try to get events. We could have two modes of the downloader perhaps to simplify the logic?